src/thumbnailer_slope_optim.cc

```diff
@@ -0,0 +1,253 @@
+
+  return kOk;
+}
+
+Thumbnailer::Status Thumbnailer::ComputeSlope(const int& ind,
+                                              const int& low_quality,
+                                              const int& high_quality,
+                                              float* const slope) {
+  frames_[ind].config.quality = low_quality;
+  int low_size;
+  float low_psnr;
+  if (GetPictureStats(frames_[ind].pic, frames_[ind].config, &low_size,
+                      &low_psnr) != kOk) {
+    return kStatsError;
+  }
+
+  frames_[ind].config.quality = high_quality;
+  int high_size;
+  float high_psnr;
+  if (GetPictureStats(frames_[ind].pic, frames_[ind].config, &high_size,
+                      &high_psnr) != kOk) {
+    return kStatsError;
+  }
+
+  *slope = (high_psnr - low_psnr) / float(high_size - low_size);
+
+  return kOk;
+}
+
+Thumbnailer::Status Thumbnailer::LossyEncodeSlopeOptim(
+    WebPData* const webp_data) {
+  // Sort frames.
+  std::sort(frames_.begin(), frames_.end(),
+            [](const FrameData& a, const FrameData& b) -> bool {
+              return a.timestamp_ms < b.timestamp_ms;
+            });
+
+  float limit_slope;
+  if (FindMedianSlope(&limit_slope) != kOk) {
+    return kSlopeOptimError;
+  }
+
+  int min_quality = minimum_lossy_quality_;
+  int max_quality = 100;
+  WebPData new_webp_data;
+  WebPDataInit(&new_webp_data);
+
+  std::queue<int> optim_list;  // Queue of frames needed to find the quality in
+                               // the next binary search loop.
+  for (int i = 0; i < frames_.size(); i++) {
+    optim_list.push(i);
+  }
+
+  // Use binary search with slope optimization to find quality values that makes
+  // the animation fit the given byte budget. The quality value for each frame
+  // can be different.
+  while (min_quality <= max_quality && !optim_list.empty()) {
+    int mid_quality = (min_quality + max_quality) / 2;
+
+    std::queue<int> new_optim_list;
+
+    while (!optim_list.empty()) {
+      int curr_frame = optim_list.front();
+      optim_list.pop();
+
+      float curr_slope;
+      if (ComputeSlope(curr_frame, min_quality, max_quality, &curr_slope) !=
+          kOk) {
+        return kSlopeOptimError;
+      }
+
+      if (frames_[curr_frame].final_quality == -1 || curr_slope > limit_slope) {
+        frames_[curr_frame].config.quality = mid_quality;
+        new_optim_list.push(curr_frame);
+      }
+    }
+
+    if (new_optim_list.empty()) break;
+
+    if (GenerateAnimationNoBudget(&new_webp_data) != kOk) {
+      return kMemoryError;
+    }
+
+    if (new_webp_data.size <= byte_budget_) {
+      while (!new_optim_list.empty()) {
+        int curr_frame = new_optim_list.front();
+        new_optim_list.pop();
+        frames_[curr_frame].final_quality = mid_quality;
+        optim_list.push(curr_frame);
+      }
+      WebPDataClear(webp_data);
+      *webp_data = new_webp_data;
+      min_quality = mid_quality + 1;
+    } else {
+      optim_list = new_optim_list;
```

Comment:
During the next pass, the quality of the frames that have `final_quality = -1` or `slope(min_quality, max_quality) >= limit_slope`  will be changed. When we're lowering the quality, it means the frames in `new_optim_list` did not get changed, so the `optim_list` in the next pass will be identical to the previous `new_optim_list`. Also, when we lowering the `max_quality` the new slope for each frame will be clearly higher than the older one so it will pass the condition at line 146.